File: typescript/src/swagger20-server.ts
import { swagger20ConfigRequired } from "./swagger20-context.js";
import {
  arrayMember,
  stringMember,
  type Swagger20Document,
  type Swagger20ResolvedOperation,
} from "./swagger20-model.js";

/** Resolves the exact effective Swagger 2.0 target base without slash repair. */
export function resolveSwagger20Server(
  document: Swagger20Document,
  operation: Swagger20ResolvedOperation,
  configured?: string,
  schemeIndex?: number,
): string {
  if (!operation.path.startsWith("/")) throw new Error(`Swagger 2.0 Paths key ${JSON.stringify(operation.path)} must begin with /`);
  if (configured !== undefined) {
    if (schemeIndex !== undefined) throw new Error("configuration.server cannot combine a complete URL with a scheme index");
    validateConfiguredServer(configured);
    return configured;
  }

  const host = stringMember(document.root, "host");
  if (host.present && (!host.valid || host.value === "" || !swagger20Host(host.value!))) {
    throw new Error("Swagger 2.0 host must contain only an authority host and optional port");
  }
  const basePath = stringMember(document.root, "basePath");
  if (basePath.present && (!basePath.valid || basePath.value === "" || !basePath.value!.startsWith("/") || /[?#]/u.test(basePath.value!))) {
    throw new Error("Swagger 2.0 basePath must be a nonempty absolute path without query or fragment");
  }
  const retrieval = document.entry.retrieval ?? document.entry.requested;
  const { usable, authored } = effectiveSchemes(document, operation, retrieval);
  let selected: string;
  if (schemeIndex !== undefined) {
    if (!Number.isSafeInteger(schemeIndex) || schemeIndex < 0 || schemeIndex >= authored.length) {
      throw new Error(`Swagger 2.0 server scheme index ${schemeIndex} is outside the effective scheme list`);
    }
    selected = authored[schemeIndex]!;
    if (selected !== "http" && selected !== "https") {
      throw new Error(`Swagger 2.0 effective scheme ${JSON.stringify(selected)} is unusable for HTTP execution`);
    }
  } else {
    if (usable.length === 0) throw new Error("Swagger 2.0 target has no usable http or https scheme");
    if (usable.length !== 1) throw swagger20ConfigRequired("server", "");
    selected = usable[0]!;
  }
  let effectiveHost = host.value;
  if (!host.present) {
    if (!retrieval) throw new Error("Swagger 2.0 target omits host without a document retrieval authority");
    let parsed: URL;
    try { parsed = new URL(retrieval); }
    catch (error: unknown) { throw new Error("Swagger 2.0 document retrieval URI is invalid", { cause: error }); }
    if (parsed.host === "") throw new Error("Swagger 2.0 target omits host without a document retrieval authority");
    effectiveHost = parsed.host;
  }
  return `${selected}://${effectiveHost}${basePath.present ? basePath.value : ""}`;
}

function effectiveSchemes(
  document: Swagger20Document,
  operation: Swagger20ResolvedOperation,
  retrieval: string | undefined,
): { usable: string[]; authored: string[] } {
  let member = arrayMember(operation.raw, "schemes");
  if (!member.present) member = arrayMember(document.root, "schemes");
  let authored: string[];
  if (!member.present) {
    if (!retrieval) throw new Error("Swagger 2.0 target omits schemes without a document retrieval scheme");
    let parsed: URL;
    try { parsed = new URL(retrieval); }
    catch (error: unknown) { throw new Error("Swagger 2.0 document retrieval URI is invalid", { cause: error }); }
    const inherited = parsed.protocol.slice(0, -1).toLowerCase();
    if (!["http", "https", "ws", "wss"].includes(inherited)) {
      throw new Error(`Swagger 2.0 retrieval scheme ${JSON.stringify(inherited)} is not an effective HTTP or WebSocket scheme`);
    }
    authored = [inherited];
  } else {
    if (!member.valid || member.value!.length === 0) throw new Error("Swagger 2.0 effective schemes must be a nonempty array");
    authored = member.value!.map((value, index) => {
      if (typeof value !== "string") throw new Error(`Swagger 2.0 effective scheme ${index} is not a string`);
      if (!["http", "https", "ws", "wss"].includes(value)) {
        throw new Error(`Swagger 2.0 effective scheme ${JSON.stringify(value)} is outside the closed scheme set`);
      }
      return value;
    });
  }
  return { authored, usable: authored.filter((scheme) => scheme === "http" || scheme === "https") };
}

function swagger20Host(value: string): boolean {
  if (/[/?#@]/u.test(value) || value.includes("://")) return false;
  try {
    const parsed = new URL(`http://${value}`);
    return parsed.host === value && parsed.username === "" && parsed.password === "" && parsed.pathname === "/"
      && parsed.search === "" && parsed.hash === "";
  } catch { return false; }
}

function validateConfiguredServer(value: string): void {
  if (value === "" || /[\r\n]/u.test(value)) throw new Error("Swagger 2.0 consumer server override is not an absolute target URL");
  let parsed: URL;
  try { parsed = new URL(value); }
  catch (error: unknown) { throw new Error("Swagger 2.0 consumer server override is not an absolute target URL", { cause: error }); }
  if ((parsed.protocol !== "http:" && parsed.protocol !== "https:") || parsed.host === "" || parsed.search !== "" || parsed.hash !== "") {
    throw new Error("Swagger 2.0 consumer server override is not a complete HTTP target URL");
  }
}

File: go/internal/runtime/swagger20_server.go
package openapiclient

import (
	"fmt"
	"net/url"
	"strings"
)

// resolveSwagger20Server builds the exact Swagger 2.0 target base. A complete
// configured URL replaces the artifact scheme, authority, and basePath; an
// index selects one authored effective scheme when more than one HTTP scheme
// remains usable.
func resolveSwagger20Server(document *Swagger20Document, operation swagger20Operation, configured string, schemeIndex *int) (string, error) {
	if document == nil || document.entry == nil {
		return "", fmt.Errorf("Swagger 2.0 document has no entry resource")
	}
	if !strings.HasPrefix(operation.path, "/") {
		return "", fmt.Errorf("Swagger 2.0 Paths key %q must begin with /", operation.path)
	}
	if configured != "" {
		if schemeIndex != nil {
			return "", fmt.Errorf("configuration.server cannot combine a complete URL with a scheme index")
		}
		if err := validateSwagger20ConfiguredServer(configured); err != nil {
			return "", err
		}
		return configured, nil
	}

	root := document.root
	host := root.string("host")
	if host.present && (!host.valid || host.value == "" || !swagger20Host(host.value)) {
		return "", fmt.Errorf("Swagger 2.0 host must contain only an authority host and optional port")
	}
	basePath := root.string("basePath")
	if basePath.present && (!basePath.valid || basePath.value == "" || !strings.HasPrefix(basePath.value, "/") || strings.ContainsAny(basePath.value, "?#")) {
		return "", fmt.Errorf("Swagger 2.0 basePath must be a nonempty absolute path without query or fragment")
	}

	retrieval := document.entry.base()
	effectiveSchemes, authored, err := swagger20EffectiveSchemes(root, operation, retrieval)
	if err != nil {
		return "", err
	}
	selected := ""
	if schemeIndex != nil {
		if *schemeIndex < 0 || *schemeIndex >= len(authored) {
			return "", fmt.Errorf("Swagger 2.0 server scheme index %d is outside the effective scheme list", *schemeIndex)
		}
		candidate := authored[*schemeIndex]
		if candidate != "http" && candidate != "https" {
			return "", fmt.Errorf("Swagger 2.0 effective scheme %q is unusable for HTTP execution", candidate)
		}
		selected = candidate
	} else {
		if len(effectiveSchemes) != 1 {
			if len(effectiveSchemes) == 0 {
				return "", fmt.Errorf("Swagger 2.0 target has no usable http or https scheme")
			}
			return "", fmt.Errorf("Swagger 2.0 target has %d usable schemes; configuration.server must select one", len(effectiveSchemes))
		}
		selected = effectiveSchemes[0]
	}

	effectiveHost := host.value
	if !host.present {
		if retrieval == nil || retrieval.Host == "" {
			return "", fmt.Errorf("Swagger 2.0 target omits host without a document retrieval authority")
		}
		effectiveHost = retrieval.Host
	}
	// An absent basePath means the API is served directly under the host, so it
	// contributes no path segment: the Paths key appends straight to the
	// authority with no synthetic "/" and no "//" composition. An authored
	// basePath keeps its exact bytes except that a lone root slash is the same
	// root boundary as an omitted basePath. Joining it to a Paths key must not
	// manufacture a double slash.
	effectiveBasePath := ""
	if basePath.present && basePath.value != "/" {
		effectiveBasePath = basePath.value
	}
	return selected + "://" + effectiveHost + effectiveBasePath, nil
}

func swagger20EffectiveSchemes(root swagger20Object, operation swagger20Operation, retrieval *url.URL) (usable, authored []string, err error) {
	member := root.array("schemes")
	if operationMember := operation.raw.array("schemes"); operationMember.present {
		member = operationMember
	}
	if !member.present {
		if retrieval == nil || retrieval.Scheme == "" {
			return nil, nil, fmt.Errorf("Swagger 2.0 target omits schemes without a document retrieval scheme")
		}
		inherited := strings.ToLower(retrieval.Scheme)
		if inherited != "http" && inherited != "https" && inherited != "ws" && inherited != "wss" {
			return nil, nil, fmt.Errorf("Swagger 2.0 retrieval scheme %q is not an effective HTTP or WebSocket scheme", retrieval.Scheme)
		}
		authored = []string{inherited}
	} else {
		if !member.valid || len(member.value) == 0 {
			return nil, nil, fmt.Errorf("Swagger 2.0 effective schemes must be a nonempty array")
		}
		for index, value := range member.value {
			scheme, ok := value.(string)
			if !ok {
				return nil, nil, fmt.Errorf("Swagger 2.0 effective scheme %d is not a string", index)
			}
			if scheme != "http" && scheme != "https" && scheme != "ws" && scheme != "wss" {
				// The defect belongs to this scheme alternative. It is removed
				// before selection; a valid sibling remains usable.
				continue
			}
			authored = append(authored, scheme)
		}
	}
	for _, scheme := range authored {
		if scheme == "http" || scheme == "https" {
			usable = append(usable, scheme)
		}
	}
	return usable, authored, nil
}

func swagger20Host(value string) bool {
	if strings.ContainsAny(value, "/?#@") || strings.Contains(value, "://") {
		return false
	}
	parsed, err := url.Parse("http://" + value)
	return err == nil && parsed.Host == value && parsed.User == nil && parsed.Path == "" && parsed.RawQuery == "" && parsed.Fragment == ""
}

func validateSwagger20ConfiguredServer(value string) error {
	if err := validateServerBaseSpelling(value); err != nil {
		return err
	}
	if !denotesTargetBase(value) {
		return fmt.Errorf("Swagger 2.0 consumer server override is not an absolute target URL")
	}
	parsed, err := url.Parse(value)
	if err != nil {
		return err
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return fmt.Errorf("Swagger 2.0 consumer server scheme %q is unusable", parsed.Scheme)
	}
	if parsed.Host == "" {
		return fmt.Errorf("Swagger 2.0 consumer server override has no authority")
	}
	if parsed.User != nil {
		return fmt.Errorf("Swagger 2.0 consumer server override contains forbidden userinfo")
	}
	return nil
}
